File: array/medium/search_in_rotated_sorted_array_33.go
package medium

func search(nums []int, target int) int {
	min, max := 0, len(nums)-1

	for min <= max {
		middle := (min + max) / 2

		if nums[middle] == target {
			return middle
		} else if nums[middle] >= nums[min] {
			if nums[min] <= target && target <= nums[middle] {
				max = middle - 1
			} else {
				min = middle + 1
			}
		} else {
			if nums[middle] <= target && target <= nums[max] {
				min = middle + 1
			} else {
				max = middle - 1
			}
		}
	}
	return -1
}

File: array/medium/java_solutions/SearchInRotatedSortedArray_33.java
package array.medium.java_solutions;

public class SearchInRotatedSortedArray_33 {

    public int search(int[] nums, int target) {
        int length = nums.length;
        int min = 0;
        int max = length - 1;
        while (min <= max) {
            int mid = (max + min) / 2;
            if (nums[mid] == target) {
                return mid;
            }
            if (nums[min] <= nums[mid]) {
                if (nums[min] <= target && target <= nums[mid]) {
                    max = mid - 1;
                } else {
                    min = mid + 1;
                }
            } else {
                if (nums[mid] <= target && target <= nums[max]) {
                    min = mid + 1;
                } else {
                    max = mid - 1;
                }
            }
        }
        return -1;
    }
}
